### PR TITLE
GCE Discovery

### PR DESCRIPTION
**Pull request by theMagicalKarp**
### What
- set up a plugin module for routing request to specific plugins
- created a GCE module for handling GCE specific tasks

Currently lighthouse can detect other vms running inside a GCE project.  Specifically it detects vms that share the same project as itself. 
### Endpoints
- `/plugins/gce/vms` lists all the vms that lighthouse has discovered
- `/plugins/gce/vms/find` tells lighthouse to kick off a task to search for local vms, this requires an authorization code from google to properly run
- `/plugins/gce/authorize` redirects the user to googles gce two-factor login, which in turn once loged in, redirects the user back to our `/plugins/gce/vms/find` with an authorization code
### DiscoveredVM Struct
- **Name**  the name of the vm
- **Address** the internal private ip address of the vm, which is only exposed to other vms sharing the project
- **CanAccessDocker** this tells us whether or not we can access docker on the vm privately, we can detect this by hitting the `/v1/_ping/` endpoint on port 2375.
### Notes

Right now I'm storing data aggregated from the GCE plugin in a JSON file. I'll go through and change it later when we decide on a database.

### CODE REVIEW

**Comment by theMagicalKarp** [comment]
@ngmiller @chfogerty @zach-taylor @cmbrose 
There is some overlap with pr https://github.com/ngmiller/lighthouse/pull/6.  I should just have to add this line, [#7 (diff)](https://github.com/ngmiller/lighthouse/blob/gce-discovery/backend/app.go#L15), to this file https://github.com/ngmiller/lighthouse/pull/6#discussion-diff-18739638R23. Once we have those changes merged.

**Comment by nickmonad** [comment]
+1

File: backend/app.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    "github.com/gorilla/mux"
+)
+
+
+func main() {
+    r := mux.NewRouter()
+
+    plugins.Handle(r.PathPrefix("/plugins").Subrouter())
+
+    http.Handle("/", r)
```
Comment (nickmonad) [inline backend/app.go]: @theMagicalKarp Could we extend the routing here to handle `/static/` for front-end assets? (https://trello.com/c/HQuz3cI9/42-remove-html-rendering-from-back-end-and-set-it-up-to-host-files-from-static)

Comment (nickmonad) [inline backend/app.go]: Well, maybe we should just keep that as a separate PR. I don't think it will be too difficult :)

**Comment by theMagicalKarp** [comment]
Jeeze Nick! What are ya doin! That's not in the scope of this ticket! :stuck_out_tongue:

**Comment by zach-taylor** [comment]
+1

**Comment by nickmonad** [comment]
+1

**Comment by nickmonad** [comment]
@chfogerty @cmbrose Any more input here?

File: Dockerfile [hunk 2]
```diff
@@ -63,6 +65,9 @@ RUN git clone git@github.com:lighthouse/lighthouse-client.git
 # Backend deps
 WORKDIR /data/lighthouse
 RUN go get github.com/fsouza/go-dockerclient
```
Comment (cmbrose) [inline Dockerfile:67]: We're not using this anymore as far as I know

File: backend/app.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import (
+    "./plugins"
+
+    "net/http"
+
+    "github.com/gorilla/mux"
+)
+
+
+func main() {
```
Comment (cmbrose) [inline backend/app.go:29]: Is this going to fight the main in backend.go?

Do we need to merge them?

**Comment by cmbrose** [comment]
Made a couple notes.  Otherwise +1!

File: backend/app.go (hunk 3 above)
Comment (theMagicalKarp) [inline backend/app.go:29]: When we get all the code merged I'll just put `plugins.Handle(r.PathPrefix("/plugins").Subrouter())` inside `bakcend.go` and delete this file.

File: Dockerfile (hunk 2 above)
Comment (theMagicalKarp) [inline Dockerfile:67]: Yup, this is just some code that got left in from a previous concept.  We can get rid of it in a "clean up" branch.

**Comment by chfogerty** [comment]
+1
